nemo/utils/exp_manager.py

```diff
@@ -106,124 +239,172 @@ def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictCo
             "You are running multi-gpu without ddp.Please note that this is not tested in NeMo and could result in "
             "errors."
         )
-    # Default root_dir to ./NeMo_experiments if None was passed
-    _root_dir = cfg.root_dir
-    if cfg.root_dir is None:
-        _root_dir = str(Path.cwd() / 'NeMo_experiments')
-    # If the user has already defined a logger for the trainer, use the logger defaults for logging directory
-    if trainer.logger is not None:
-        if trainer.logger.save_dir:
-            if cfg.root_dir:
-                raise ValueError(
-                    "The pytorch lightning trainer that was passed to exp_manager contained a logger, the logger's "
-                    f"save_dir was not None, and root_dir ({cfg.root_dir}) was not None. If trainer.logger.save_dir "
-                    "exists, exp_manager will use trainer.logger.save_dir as the logging directory and root_dir "
-                    "must be None."
-                )
-            _root_dir = trainer.logger.save_dir
-        if cfg.name:
-            raise ValueError(
-                "The pytorch lightning trainer that was passed to exp_manager contained a logger, and name "
-                f"({cfg.name}) was also passed to exp_manager. If the trainer contains a logger, exp_manager will use "
-                "trainer.logger.name, and name that is passed to exp_manager must be None."
-            )
-        if cfg.create_tensorboard_logger:
+
+
+def check_resume(trainer: 'pytorch_lightning.Trainer', log_dir: str, resume_past_end: bool = False):
+    """Checks that resume=True was used correctly with the arguments pass to exp_manager. Sets
+    trainer.resume_from_checkpoint as necessary.
+
+    Returns:
+        log_dir (Path): the log_dir
+        exp_dir (str): the base exp_dir without name nor version
+        name (str): The name of the experiment
+        version (str): The version of the experiment
+
+    Raises:
+        LoggerMisconfigurationError: If trainer is incompatible with arguments
+        ValueError: If resume is True and checkpoints could not be find.
+    """
+    if not log_dir:
+        raise ValueError(f"Resuming requires the log_dir {log_dir} to be passed to exp_manager")
+
+    checkpoint_dir = Path(Path(log_dir) / "checkpoints")
+    checkpoint = None
+    end_checkpoints = list(checkpoint_dir.glob("*end.ckpt"))
+    last_checkpoints = list(checkpoint_dir.glob("*last.ckpt"))
+    if not checkpoint_dir.exists():
+        raise NotFoundError(f"There was no checkpoint folder at checkpoint_dir :{checkpoint_dir}. Cannot resume.")
```

Comment:
I don't understand why both @blisc and @tkornuta-nvidia think that hard crashing is a good idea if checkpoints are not found. This artificially causes friction when a user wants to perform multiple training runs. 

What is the rationale between forcing users to take extra steps like this ? It's definitely not a design decision. Unlike Nemo 1, where we did not have an explicit "resume" flag, here this code path only ever executed if the user voluntarily choses to set resume = True. The user has already agreed to the contract that resume behaviour is expected and necessary so then why do we want to artificially cause friction when it's unnecessary?

Reply:
This design inherits from OpenSeq2Seq where users had to explicitly pass in a resume flag for an experiment to resume. This design in NeMo 0.x is one of the ones I believe should be improved upon in NeMo 1.0.

If a user wants to resume training, they should only be able to resume if they have a checkpoint. If they set resume=True and there is no checkpoint, how is this anything but an error? Like you said, the user has agreed that resume behaviour is expected and necessary, but without a checkpoint, how can the software do what the user expects?

Reply:
Alright then the `resume` flag should be renamed `resume_if_exists` and that solves the usability issue? 

Resume if the checkpoint exists when set to true, will resume training when checkpoing is there and train from scratch if not there. If set to false, then no matter checkpoint exists or not, it will train from scratch. 

Does this sound like a more reasonable flag?

Reply:
I think @titu1994 concern is from the scenario where it would be convenient to issue exactly the same train command, with same checkpoint dir (e.g. automatically when cluster has wall time limits). This would resume for runs n>1 but for run n=1 will fail because checkpoints haven't been created yet. 

But this: "If a user wants to resume training, they should only be able to resume if they have a checkpoint." sounds reasonable too. And I would add that silently resuming (e.g. without explicit --resume flag did lead to errors before).

I think having both "--resume" and "--resume_if_exists" sounds like a good solution. The only difference being that "--resume" will fail if checkpoints do not exists and "--resume_if_exists" will not. @blisc what do you think about that?

Reply:
That's a fair point, that not having resume flag (and still resuming) is definitely not correct behaviour. 

Adding two flags, which do nearly the same thing, however is an added cognitive burdon - both for the exp manager and for the user. 

Case where you would use resume and not resume_if_exists:
1) Model checkpoint downloaded from NGC
2) Continue training for more epochs on same dataset (but have 1 checkpoint at the minimum from somewhere)

Case where you would use resume_if_exists but not resume:
(1) same as above (1)
(2) Same as (2) above
(3) Same as (2) above, but now you don't have a checkpoint even at the beginning (loop train case from scratch)

Resume if exists is a strict superset of resume, with the semantic that you want explicitly attack the resume training behaviour, but fall back to from scratch training if it is not possible to load the checkpoint. 

Of course, without either flag, it should completely ignore previous checkpoints.

Reply:
Let me rename resume to resume_if_exists and add a new flag resume_ignore_errors that will avoid raising exceptions.
`resume_ignore_errors` will default to False and raise exceptions. Users can override this to True and avoid erroring out